optuna/pruners/hyperband.py

```diff
@@ -0,0 +1,155 @@
+            reduction_factor=3,
+            n_brackets=4,
+            min_early_stopping_rate_low=0
+    ):
+        # type: (int, int, int, int) -> None
+
+        self._pruners = []  # type: List[SuccessiveHalvingPruner]
+        self._reduction_factor = reduction_factor
+        self._resource_budget = 0
+        self._n_brackets = n_brackets
+        self._bracket_resource_budgets = []  # type: List[int]
+
+        _logger.debug('Hyperband has {} brackets'.format(self._n_brackets))
+
+        for i in range(n_brackets):
+            bracket_resource_budget = self._calc_bracket_resource_budget(i, n_brackets)
+            self._resource_budget += bracket_resource_budget
+            self._bracket_resource_budgets.append(bracket_resource_budget)
+
+            # N.B. (crcrpar): `min_early_stopping_rate` has the information of `bracket_index`.
+            min_early_stopping_rate = min_early_stopping_rate_low + i
+
+            _logger.debug(
+                '{}th bracket has minimum early stopping rate of {}'.format(
+                    i, min_early_stopping_rate))
+
+            pruner = SuccessiveHalvingPruner(
+                min_resource=min_resource,
+                reduction_factor=reduction_factor,
+                min_early_stopping_rate=min_early_stopping_rate,
+            )
+            self._pruners.append(pruner)
+
+    def prune(self, study, trial):
+        # type: (Study, structs.FrozenTrial) -> bool
+
+        i = self._get_bracket_id(study, trial)
+        _logger.debug('{}th bracket is selected'.format(i))
+        bracket_study = self._create_bracket_study(study, i)
+        return self._pruners[i].prune(bracket_study, trial)
+
+    def _calc_bracket_resource_budget(self, pruner_index, n_brackets):
+        # type: (int, int) -> int
+
+        n = self._reduction_factor ** (n_brackets - 1)
+        budget = n
+        for i in range(pruner_index, n_brackets - 1):
+            budget += n / 2
+        return budget
+
+    def _get_bracket_id(self, study, trial):
+        # type: (Study, structs.FrozenTrial) -> int
+        """Computes the index of bracket for a trial of ``trial_number``.
+
+        The index of a bracket is noted as :math:`s` in
+        `Hyperband paper <http://www.jmlr.org/papers/volume18/16-558/16-558.pdf>`_.
+        """
+
+        n = hash('{}_{}'.format(study.study_name, trial.number)) % self._resource_budget
+        for i in range(self._n_brackets):
+            n -= self._bracket_resource_budgets[i]
+            if n < 0:
+                return i
+
+        assert False
+
+    def _create_bracket_study(self, study, bracket_index):
+        # type: (Study, int) -> Study
+
+        from optuna.study import Study
+
+        # N.B. This class is assumed to be passed to
+        # `SuccessiveHalvingPruner.prune` in which `get_trials`,
+        # `direction`, and `storage` are used.
+        # But for safety, prohibit the other attributes explicitly.
+        class _BracketStudy(Study):
```

Comment:
**FYI** I needed to apply the following diffs to make this study class work:
```diff
diff --git a/optuna/pruners/hyperband.py b/optuna/pruners/hyperband.py
index e5a225ad..90d15df7 100644
--- a/optuna/pruners/hyperband.py
+++ b/optuna/pruners/hyperband.py
@@ -122,7 +122,7 @@ class HyperbandPruner(BasePruner):
         # But for safety, prohibit the other attributes explicitly.
         class _BracketStudy(Study):
 
-            _VALID_ATTRS = ('get_trials', 'direction', '_storage')
+            _VALID_ATTRS = ('get_trials', 'direction', '_storage', '_study_id', 'pruner', 'study_name', '_bracket_id')
 
             def __init__(self, study, bracket_id):
                 # type: (Study, int) -> None
@@ -150,6 +150,6 @@ class HyperbandPruner(BasePruner):
                 if attr_name not in _BracketStudy._VALID_ATTRS:
                     raise NotImplementedError
                 else:
-                    return getattr(self, attr_name)
+                    return object.__getattribute__(self, attr_name)
 
         return _BracketStudy(study, bracket_index)
```

Reply:
😔

Thank you very much. I should've written more tests.